export_as_locals.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2021 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terramate
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/madlambda/spells/errutil"
+	"github.com/mineiros-io/terramate/config"
+	"github.com/mineiros-io/terramate/hcl"
+	"github.com/mineiros-io/terramate/hcl/eval"
+	"github.com/zclconf/go-cty/cty"
+)
+
+const ErrExportAsLocalsRedefined errutil.Error = "export_as_locals attribute redefined"
+
+// ExportAsLocals represents information obtained by parsing and evaluating
+// export_as_locals blocks.
+type ExportAsLocals struct {
+	attributes map[string]cty.Value
+}
+
+// LoadStackExportAsLocals loads from the file system all export_as_locals for
+// a given stack. It will navigate the file system from the stack dir until
+// it reaches rootdir, loading export_as_locals and merging them appropriately.
+//
+// More specific definitions (closer or at the stack) have precedence over
+// less specific ones (closer or at the root dir).
+//
+// Metadata and globals for the stack are used on the evaluation of the
+// export_as_locals blocks.
+//
+// The rootdir MUST be an absolute path.
+func LoadStackExportAsLocals(rootdir string, sm StackMetadata, g *Globals) (ExportAsLocals, error) {
+	if !filepath.IsAbs(rootdir) {
+		return ExportAsLocals{}, fmt.Errorf("%q must be an absolute path", rootdir)
+	}
+
+	unEvalExport, err := loadStackExportAsLocals(rootdir, sm.Path)
+	if err != nil {
+		return ExportAsLocals{}, err
+	}
+	return unEvalExport.eval(sm, g)
+}
+
+func (e ExportAsLocals) Attributes() map[string]cty.Value {
+	attrcopy := map[string]cty.Value{}
+	for k, v := range e.attributes {
+		attrcopy[k] = v
+	}
+	return attrcopy
+}
+
+func loadStackExportAsLocals(rootdir string, cfgdir string) (unEvalExportAsLocals, error) {
+	cfgpath := filepath.Join(rootdir, cfgdir, config.Filename)
+	blocks, err := hcl.ParseExportAsLocalsBlocks(cfgpath)
+
+	if os.IsNotExist(err) {
+		parentcfg, ok := parentDir(cfgdir)
+		if !ok {
+			return newUnEvalExportAsLocals(), nil
+		}
+		return loadStackExportAsLocals(rootdir, parentcfg)
+	}
+
+	if err != nil {
+		return unEvalExportAsLocals{}, err
+	}
+
+	exportLocals := newUnEvalExportAsLocals()
```

Comment:
`newUnEvalExportAsLocals()` again newUnEval is strange and AsLocal looks like converting something. Of course the context of the file is `export_as_locals` but it doesn't mean we need to embed this everywhere.

Reply:
Man adding the "As" everywhere was the dumbest idea I ever had in my life..honestly..WTF. +1 on improving.